crates/wasmparser/src/readers/core/types.rs

```diff
@@ -340,10 +341,27 @@ impl RefType {
     /// Returns `None` when the type index is beyond this crate's implementation
     /// limits and therefore is not representable.
     pub const fn typed_func(nullable: bool, index: u32) -> Option<Self> {
+        Self::indexed(nullable, HeapType::Func, index)
+    }
+
+    /// Create a reference to a user defined type at the given index.
+    ///
+    /// Returns `None` when the type index is beyond this crate's implementation
+    /// limits and therefore is not representable, or when the heap type is not
+    /// a typed array, struct or function.
+    pub const fn indexed(nullable: bool, heap_type: HeapType, index: u32) -> Option<Self> {
+        let kind = match heap_type {
+            HeapType::Func => Self::FUNC_KIND,
+            HeapType::Array => Self::ARRAY_KIND,
+            HeapType::Struct => Self::STRUCT_KIND,
+            _ => {
+                return None;
+            }
```

Comment:
Perhaps? I'm honestly not sure. I find it really difficult to review piecemeal incremental patches like this since I don't have the whole picture in my head. I forget what the previous version of this function now but it felt wrong to pass in `TypedFunc` but get out an indexed array (or something like that). Even now returning `None` with a catch-all `_` feels wrong here since that should probably indicate a programmer error than anything else. Additionally this is where I'm not sure that the design of `HeapType` makes sense to take as an argument here because it already has an indexed variant but it's explicitly not used here.

Overall I'm not sure how best to leave this. If it sticks with `indexed(...)` then I think it should take the `Self::*_KIND` as an argument rather than `HeapType`. Otherwise I'm not sure how best to organize this.

Reply:
Done the refactoring with separate functions for each kind of indexed ref.